core/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala

```diff
@@ -198,6 +198,78 @@ class QualificationAppInfo(
     }
   }
 
+  private def singleExecMatch(execInfo: ExecInfo): (Seq[(Int, ExecInfo)], Option[ExecInfo]) = {
+    val associatedStages = {
+      if (execInfo.stages.size > 1) {
+        execInfo.stages.toSeq
+      } else if (execInfo.stages.size < 1) {
+        // we don't know what stage its in or its duration
+        logDebug(s"No stage associated with ${execInfo.exec} " +
+          s"so speedup factor isn't applied anywhere.")
+        Seq.empty
+      } else {
+        Seq(execInfo.stages.head)
+      }
+    }
+    if (associatedStages.nonEmpty) {
+      (associatedStages.map((_, execInfo)), None)
+    } else {
+      (Seq.empty, Some(execInfo))
+    }
+  }
+
+  private def doubleExecMatch(neighbor: ExecInfo, execInfo: ExecInfo): (
+    Seq[(Int, ExecInfo)], Option[ExecInfo]) = {
+    val associatedStages = {
+      if (execInfo.stages.size > 1) {
+        execInfo.stages.toSeq
+      } else if (execInfo.stages.size < 1) {
+        if (neighbor.stages.size >= 1) {
+          neighbor.stages.headOption.toSeq
+        } else {
+          // we don't know what stage its in or its duration
+          logDebug(s"No stage associated with ${execInfo.exec} " +
+            s"so speedup factor isn't applied anywhere.")
+          Seq.empty
+        }
+      } else {
+        Seq(execInfo.stages.head)
+      }
+    }
+    if (associatedStages.nonEmpty) {
+      (associatedStages.map((_, execInfo)), None)
+    } else {
+      (Seq.empty, Some(execInfo))
+    }
+  }
+
+  private def tripleExecMatch(prev: ExecInfo, execInfo: ExecInfo, next: ExecInfo):
```

Comment:
Logic for `singleExecMatch`, `doubleExecMatch` and `tripleExecMatch` are quite similar. Can we combine them by using a seq of `ExecInfo` as neighbors?

Reply:
Consolidated it into single function. PTAL

Reply:
Thanks @nartal1